pyanaconda/dbus/interface.py

```diff
@@ -0,0 +1,323 @@
+        return inspect.ismethod(member) or inspect.isfunction(member)
+
+    def _generate_method(self, member, member_name):
+        """Generate method defined by given class member.
+
+        :param member: a method object
+        :param member_name: a name of the method
+        :return: a method element
+        """
+        method = self.xml_generator.get_method_element(member_name)
+
+        # Process the parameters.
+        for name, type_hint, direction in self._iterate_parameters(member):
+            # Create the parameter element.
+            parameter = self.xml_generator.create_parameter(name, get_dbus_type(type_hint), direction)
+            # Add the element to the method element.
+            self.xml_generator.add_child(method, parameter)
+
+        return method
+
+    def _is_property(self, member):
+        """Is the class member a DBus property?"""
+        return isinstance(member, property)
+
+    def _generate_property(self, member, member_name):
+        """Generate DBus property defined by class member.
+
+        :param member: a property object
+        :param member_name: a property name
+        :return: a property element
+        """
+        access = None
+        type_hint = None
+
+        try:
+            # Process the setter.
+            if member.fset:
+                [(_, type_hint, _)] = self._iterate_parameters(member.fset)
+                access = self.ACCESS_WRITE
+
+            # Process the getter.
+            if member.fget:
+                [(_, type_hint, _)] = self._iterate_parameters(member.fget)
+                access = self.ACCESS_READ
+
+        except ValueError:
+            raise DBusSpecificationError("Property %s has invalid parameters." % member_name)
+
+        # Property has both.
+        if member.fget and member.fset:
+            access = self.ACCESS_READWRITE
+
+        if access is None:
+            raise DBusSpecificationError("Property %s is not accessible." % member_name)
+
+        return self.xml_generator.create_property(member_name, get_dbus_type(type_hint), access)
+
+    def _is_signal(self, member):
+        """Is the class member a DBus signal?"""
+        return isinstance(member, dbus_signal)
+
+    def _generate_signal(self, member, member_name):
+        """Generate signal defined by a class member.
+
+        :param member: a dbus_signal object.
+        :param member_name: a name of the signal
+        :return: a signal element
+        """
+        element = self.xml_generator.get_signal_element(member_name)
+
+        if not member.declaration:
+            return element
+
+        for name, type_hint, direction in self._iterate_parameters(member.declaration):
+            # Only input parameters can be defined.
+            if direction == self.DIRECTION_OUT:
+                raise DBusSpecificationError("Signal %s has defined return type." % member_name)
+
+            direction = self.DIRECTION_OUT
```

Comment:
The DBus specification says, that a signal can have only OUT parameters. Since a signal emits its parameters and doesn't return anything, we should forbid the OUT parameter and transform IN parameters to OUT parameters.

Added comment.